src/Tags/Structure.php

```diff
@@ -148,6 +157,15 @@ public function toArray($tree, $parent = null, $depth = 1)
         return $pages->all();
     }
 
+    protected function remember($key, $callable)
+    {
+        return Cache::remember(
+            $key,
+            now()->addSeconds(Config::get('statamic.system.structure_cache_ttl')),
+            $callable
+        );
+    }
+
```

Comment:
(This comment applies to this entire file)

By caching these bits, it definitely speeds things up. However if you update an entry being used in a tree, it won't be reflected in the output.

We need to invalidate them somehow.

The url and absolute url is simple enough. We could listen for EntrySaved and invalidate the `"pages:urls:relative:{$ref}"` key. But for the data, that's a little more complicated because the key you've used will change depending on whether you've used the `select=""` param (`pages:augmented:{$ref}".md5(json_encode($keys))`)

Hmm. 🤔

Reply:
cache the keys (`nav_cache_keys`) then invalidate in same listener?

Reply:
I was thinking that we could bust cache based on prefix, like `pages:augmented:*`, but maybe that's me assuming too much... maybe my head was in the Redis space, because I have been using Redis for my caching needs

Reply:
Also, please see my [earlier comment](https://github.com/statamic/cms/pull/5899#issuecomment-1146099875)